iceoryx_utils/include/iceoryx_utils/internal/relocatable_pointer/base_relative_ptr.hpp

```diff
@@ -0,0 +1,184 @@
+// Copyright (c) 2019 by Robert Bosch GmbH. All rights reserved.
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_UTILS_RELOCATABLE_POINTER_BASE_RELATIVE_PTR_HPP
+#define IOX_UTILS_RELOCATABLE_POINTER_BASE_RELATIVE_PTR_HPP
+
+#include "pointer_repository.hpp"
+
+#include <cstdint>
+
+namespace iox
+{
+namespace rp
+{
+/// @brief pointer class to use when pointer and pointee are located in different shared memory segments
+/// We can have the following scenario:
+/// Pointer p is stored in segment S1 and points to object X of type T in segment S2.
+///
+/// Shared Memory   S1:  p              S2:  X
+///                      |___________________^
+/// App1            a1   b1             c1   d1
+/// App2            a2   b2             c2   d2
+///
+/// Now it is no longer true in general that both segments will be offset by the same difference in App2 and therefore
+/// relocatable pointers are no longer sufficient.
+/// Relative pointers solve this problem by incorporating the information from where they need to measure differences
+/// (i.e. relative to the given address). This requires an additional registration mechanism to be used by all
+/// applications where the start addresses and the size of all segments to be used are registered. Since these start
+/// address may differ between applications, each segment is identified by a unique id, which can be provided upon
+/// registration by the first application. In the figure, this means that the starting addresses of both segments(a1, a2
+/// and c1, c2) would have to be registered in both applications.
+/// Once this registration is done, relative pointers can be constructed from raw pointers similar to relocatable
+/// pointers.
+/// @note It should be noted that relocating a memory segment will invalidate relative pointers, i.e. relative pointers
+/// are NOT relocatable. This is because the registration mechanism cannot be automatically informed about the copy of a
+/// whole segment, such a segment would have to be registered on its own (and the original segment deregistered).
+class BaseRelativePointer
+{
+  public:
+    using id_t = uint64_t;
+    using ptr_t = void*;
+    using const_ptr_t = const void* const;
+    using offset_t = std::uintptr_t;
+
+    /// @brief constructs a BaseRelativePointer pointing to ptr in a segment identified by id
```

Comment:
"pointing to the same pointee as ptr" (may not be the best wording, but it is accurate)
This occurs multiple times.